src/box/replication.cc

```diff
@@ -206,6 +209,70 @@ replicaset_set_sync_quorum(void)
 	replication_sync_quorum_auto = replicaset.applier.connected;
 }
 
+/**
+ * Deactivates all expired anonymous replicas. They are expected to be
+ * deleted on deactivation as orphan.
+ * Returns lowest time-to-live among remaining anonymous replicas via
+ * `lowest_ttl`. If there are no anonymous replicas left or the function has
+ * failed to deactivate a replica, `replication_anon_ttl` is returned.
+ */
+static void
+replication_anon_gc(double *lowest_ttl)
+{
+	while (true) {
+		*lowest_ttl = replication_anon_ttl;
+		struct replica *expired_replica = NULL;
+		/*
+		 * Search for the first expired replica. Note that we cannot
+		 * just deactivate it and continue the iteration because a yield
+		 * will happen and the replicaset can be changed significantly.
+		 */
+		replicaset_foreach(replica) {
+			if (!replica->anon || replica_has_connections(replica))
+				continue;
+			double last_seen = relay_last_row_time(replica->relay);
+			double unused = ev_monotonic_now(loop()) - last_seen;
+			double ttl = replication_anon_ttl - unused;
+			if (ttl <= 0) {
+				expired_replica = replica;
+				break;
+			}
+			*lowest_ttl = MIN(*lowest_ttl, ttl);
+		}
+		if (expired_replica == NULL)
+			break;
+		/* Save UUID since the replica will be deleted on success. */
+		struct tt_uuid uuid = expired_replica->uuid;
+		if (replica_deactivate(expired_replica) != 0) {
```

Comment:
Most of the checks performed by `replica_deactivate` don't make sense here. I think it's better to factor out the code doing actual deactivation into a separate static function (`replica_deactivate_impl`) and make `relica_deactivate` take uuid instead of a replica object to make the Lua wrapper thin.

Reply:
Done.